pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/EmptyResponseUtils.java

```diff
@@ -0,0 +1,206 @@
+    ColumnDataType[] columnDataTypes = new ColumnDataType[numColumns];
+    int index = 0;
+    for (ExpressionContext groupByExpression : groupByExpressions) {
+      columnNames[index] = groupByExpression.toString();
+      // Use STRING column data type as default for group-by expressions
+      columnDataTypes[index] = ColumnDataType.STRING;
+      index++;
+    }
+    for (Pair<AggregationFunction, FilterContext> pair : filteredAggregationFunctions) {
+      AggregationFunction aggregationFunction = pair.getLeft();
+      columnNames[index] = AggregationFunctionUtils.getResultColumnName(aggregationFunction, pair.getRight());
+      columnDataTypes[index] = aggregationFunction.getFinalResultColumnType();
+      index++;
+    }
+    return new ResultTable(new DataSchema(columnNames, columnDataTypes), List.of());
+  }
+
+  /**
+   * Tries to fill an {@link DataSchema} when no row has been returned.
+   *
+   * Response data schema can be inaccurate (all columns set to STRING) when all segments are pruned on broker or
+   * server.
+   *
+   * Priority is:
+   * - Types from multi-stage engine validation for the given query (if allowed).
+   * - Types from schema for the given table (only applicable to identifiers).
+   * - Types from single-stage engine response (no action).
+   *
+   * Multi-stage engine schema will be available only if query compiles.
+   */
+  public static void fillEmptyResponseSchema(boolean useMSE, BrokerResponse response, TableCache tableCache,
+      Schema schema, String database, String query) {
+    Preconditions.checkState(response.getNumRowsResultSet() == 0, "Cannot fill schema for non-empty response");
+
+    DataSchema dataSchema = response.getResultTable() != null ? response.getResultTable().getDataSchema() : null;
+
+    List<RelDataTypeField> dataTypeFields = null;
+    // Turn on (with pinot.broker.use.mse.to.fill.empty.response.schema=true or query option
+    // useMSEToFillEmptyResponseSchema=true) only for clusters where no queries with huge IN clauses are expected
+    // (see https://github.com/apache/pinot/issues/15064)
+    if (useMSE) {
+      QueryEnvironment queryEnvironment = new QueryEnvironment(database, tableCache, null);
+      RelRoot root;
+      try (QueryEnvironment.CompiledQuery compiledQuery = queryEnvironment.compile(query)) {
+        root = compiledQuery.getRelRoot();
+      } catch (Exception ignored) {
+        root = null;
+      }
+      if (root != null && root.validatedRowType != null) {
+        dataTypeFields = root.validatedRowType.getFieldList();
+      }
+    }
+
+    if (dataSchema == null && dataTypeFields == null) {
+      // No schema available, nothing we can do
+      return;
+    }
+
+    if (dataSchema == null || (dataTypeFields != null && dataSchema.size() != dataTypeFields.size())) {
+      // If data schema is not available or has different number of columns than the validated row type, we use the
+      // validated row type to populate the schema.
```

Comment:
I don't follow this part about data schema having a different number of columns than the validated row type and why we're using the validated row type in this case. Could you please elaborate?

Reply:
This method is moved from `ParserUtils` to this class as this class contains the methods handling broker side empty responses. This method is used to fill the data schema for both broker side generated empty response and server side generated empty response. I believe we keep this check to ensure we won't run into `IndexOutOfBoundException` when SSE and MSE somehow doesn't compile to the same data schema.

Reply:
But we're choosing to use the MSE data schema here for an SSE query? If there's a mismatch, wouldn't we want to stick to the SSE data schema since it's a query executed using SSE?

Reply:
This is debatable, and the current behavior trusts MSE data schema more than SSE data schema when there is no row returned.
When data schema has different number of columns, we use the one from MSE; similarly when data schema has different data type, we use the one from MSE. We can probably revisit it later